### DISCUSSION TITLE
How to set up filtering by tags?

### DISCUSSION THREAD

**Discussion by nktnet1** [fuma-nama/fumadocs#818] [Q&A] [answered]
I tried the following:

![image](https://github.com/user-attachments/assets/1993a634-6479-402b-9617-f6c283b2f090)

<details close>
<summary>code</summary>

```ts
export async function GET(req: NextRequest) {
  const session = await getAuthSession();
  const pages = session
    ? source.getPages()
    : source.getPages().filter((page) => publicContentPaths.includes(page.url));
  const searchAPI = createSearchAPI('advanced', {
    indexes: pages.map((page) => ({
      id: page.url,
      title: page.data.title,
      description: page.data.description,
      structuredData: page.data.structuredData,
      tag: page.data.tag,
      url: page.url,
    })),
    // tag: true,
  });
  return searchAPI.GET(req);
}
```

</details>

After consulting the docs here:
- https://fumadocs.vercel.app/docs/headless/search/flexsearch#tag-filter

![image](https://github.com/user-attachments/assets/d3f19860-57ad-4814-bf26-208ec52f780a)

What am I missing? Thanks!

---

Current versions:
```
    "fumadocs-core": "13.4.7",
    "fumadocs-mdx": "10.0.1",
    "fumadocs-ui": "13.4.7",
```

**Comment by fuma-nama** [comment]
the `data.tag` is a custom property, you can pass anything to it, including a custom field on your frontmatter schema.

**Reply by nktnet1** [reply]
I'm trying to implement versioned docs similar to what you have here:
- https://fumadocs.vercel.app/docs/

I've been looking at the source code:
- https://github.com/fuma-nama/fumadocs/tree/dev/apps/docs

But can't seem to figure out where or how you're restricting the searches. 

Can you show me how I can extend the frontmatter schema?
![image](https://github.com/user-attachments/assets/638ae979-4bd6-40b6-b8bb-754699069005)

Basically, I have the following structure:
```
$ tree content/periods -L 1
content/periods
├── 24T3
└── 25T1
```

and I want to configure tags similar to Core/UI/MDX, where users can only search in one of those directories at a time.

Also, would I need to specify the tag in each `.md` and `.mdx` file, or is there a way to match it based those files using their parent directory as the tag?

Thank you!

**Reply by fuma-nama** [reply, marked as answer]
> would I need to specify the tag in each .md and .mdx file, or is there a way to match it based those files using their parent directory as the tag?

just choose the way you want, `page` has its file path and frontmatter data.

to extend the frontmatter, see [`schema`](https://fumadocs-dev.vercel.app/docs/mdx/configuration#extend-schema)

**Reply by nktnet1** [reply]
Ah... I think I understand now. I thought `tag` was similar to  `title` and `description` which was a metadata property of `.md` or `.mdx` file, but we could also do things like:

```ts
      tag: page.file.flattenedPath.substring(0, page.file.flattenedPath.indexOf('/')),
```

![image](https://github.com/user-attachments/assets/06485beb-c285-498c-bc2b-82eb35d8cfd7)

I couldn't find out where `flattenedPath` was documented in the docs, but was able to find it when logging things out. I think this solves my problem!

Thanks @fuma-nama ❤️!

**Reply by fuma-nama** [reply]
Also be careful, your code creates a new instance of search server for every request, you should only create two instances: one for authorised users, one for unauthorised

**Reply by nktnet1** [reply]
Thanks @fuma-nama!

I couldn't figure it out before when I was new to NextJS and fumadocs so went with this temporarily - also wasn't sure how to import "Page", but I guess I could just do:
```ts
ReturnType<typeof source.getPages>
```

I've now fixed it:

```ts
import { source } from '@/app/source';
import { createSearchAPI } from 'fumadocs-core/search/server';
import { NextRequest } from 'next/server';
import { publicContentPaths } from '@/app/(periods)/constants';
import { getAuthSession } from '../auth/[...nextauth]/auth';

const createSearchInstance = (pages: ReturnType<typeof source.getPages>) => {
  return createSearchAPI('advanced', {
    indexes: pages.map((page) => ({
      id: page.url,
      title: page.data.title,
      description: page.data.description,
      structuredData: page.data.structuredData,
      tag: page.file.flattenedPath.substring(0, page.file.flattenedPath.indexOf('/')),
      url: page.url,
    })),
    tag: true,
  });
};

const allPages = source.getPages();
const authorizedSearchAPI = createSearchInstance(allPages);
const unauthorizedSearchAPI = createSearchInstance(
  allPages.filter((page) => publicContentPaths.includes(page.url)),
);

export async function GET(req: NextRequest) {
  const session = await getAuthSession();
  const searchAPI = session ? authorizedSearchAPI : unauthorizedSearchAPI;
  return searchAPI.GET(req);
}
```

Thanks for pointing this out - yep, I definitely do not want to be doing that on every search 😹!

**Reply by fuma-nama** [reply]
btw `InferPageType` is what you need. I'll document it later

**Reply by nktnet1** [reply]
@fuma-nama  there's currently a type error with this code after upgrading to fumadocs v14.0.1:

```ts
import { NextRequest } from 'next/server';

export async function GET(req: NextRequest) {
  const session = await auth();
  const searchAPI = session ? authorizedSearchAPI : unauthorizedSearchAPI;

  // Error here
  return searchAPI.GET(req);
}
```

The error occurs on both NextJS v14.2.16 and NextJS v15.0.1 (but not v15.0.0) - e.g.
```
// Next 14.2.16 TS error

Argument of type 'import("/home/nktnet/work/unsw/comp1531-website/node_modules/.pnpm/next@14.2.16_@babel+core@7.25.9_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/next/dist/server/web/spec-extension/request").NextRequest' is not assignable to parameter of type 'import("/home/nktnet/work/unsw/comp1531-website/node_modules/.pnpm/next@15.0.0_@babel+core@7.25.9_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/next/dist/server/web/spec-extension/request").NextRequest'.
  Property '[INTERNALS]' is missing in type 'import("/home/nktnet/work/unsw/comp1531-website/node_modules/.pnpm/next@14.2.16_@babel+core@7.25.9_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/next/dist/server/web/spec-extension/request").NextRequest' but required in type 'import("/home/nktnet/work/unsw/comp1531-website/node_modules/.pnpm/next@15.0.0_@babel+core@7.25.9_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/next/dist/server/web/spec-extension/request").NextRequest'.ts(2345)
request.d.ts(11, 5): '[INTERNALS]' is declared here.
```

```
// Next 15.0.1 TS error

Argument of type 'import("/home/nktnet/work/unsw/comp1531-website/node_modules/.pnpm/next@15.0.1_@babel+core@7.25.9_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/next/dist/server/web/spec-extension/request").NextRequest' is not assignable to parameter of type 'import("/home/nktnet/work/unsw/comp1531-website/node_modules/.pnpm/next@15.0.0_@babel+core@7.25.9_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/next/dist/server/web/spec-extension/request").NextRequest'.
  Property '[INTERNALS]' is missing in type 'import("/home/nktnet/work/unsw/comp1531-website/node_modules/.pnpm/next@15.0.1_@babel+core@7.25.9_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/next/dist/server/web/spec-extension/request").NextRequest' but required in type 'import("/home/nktnet/work/unsw/comp1531-website/node_modules/.pnpm/next@15.0.0_@babel+core@7.25.9_react-dom@18.3.1_react@18.3.1__react@18.3.1/node_modules/next/dist/server/web/spec-extension/request").NextRequest'.ts(2345)
request.d.ts(11, 5): '[INTERNALS]' is declared here.
```


The workaround currently is to use some TypeScript hacks
```ts
export async function GET(req: Parameters<ReturnType<typeof createSearchInstance>['GET']>[0]) {
  const session = await auth();
  const searchAPI = session ? authorizedSearchAPI : unauthorizedSearchAPI;
  return searchAPI.GET(req);
}
```
i.e. replacing the `NextRequest` type with `Parameters<ReturnType<typeof createSearchInstance>['GET']>[0]`

The only occurrence of this issue that I could find online was:
- https://docs.arcjet.com/troubleshooting/#property-internals-is-missing-in-type

**Reply by fuma-nama** [reply]
You may need to cast it to Next Request manually, not sure if it's a Next.js bug

**Reply by nktnet1** [reply]
Yeah casting doesn't work either, because "NextRequest" imported from "next/server" is of a different version to "NextRequest" used internally by fumadocs - unless there's a way to import the "NextRequest" used by fumadocs directly?

Definitely a version mismatch for the "NextRequest" type I think.

**Reply by fuma-nama** [reply]
Feels weird, have to investigate later

**Reply by fuma-nama** [reply]
I made a change which may fix your issue. You can cast the type for now to avoid errors
